File: framework/java/src/main/java/soda/web/BaseHandler.java
package soda.web;

import java.io.*;

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;

import soda.web.resp.Response;
import soda.web.resp.ResponseFactory;

public abstract class BaseHandler implements HttpHandler {

	protected Response doGet(HttpExchange exchange) throws Exception {
		throw new RuntimeException("not implemented");
	}

	protected Response doPost(HttpExchange exchange) throws Exception {
		throw new RuntimeException("not implemented");
	}
	
	@Override
	public void handle(HttpExchange exchange) throws IOException {
		var method = exchange.getRequestMethod();
		var uri = exchange.getRequestURI();

		Response resp = null;
		try {
			long startNano = System.nanoTime();
			
			if (method.equals("GET")) {
				resp = doGet(exchange);
			} else if (method.equals("POST")) {
				resp = doPost(exchange);
			} else {
				throw new RuntimeException("method " + method + " not supported");
			}

			long endNano = System.nanoTime();
			double elapseMs = (endNano - startNano) / 1e6;
			int code = resp.getHttpCode();
			Logger.infof("%s %s %d %f", method, uri, code, elapseMs);
		} catch (ServiceException ex) {
			String message = String.format("%s %s %d business error", method, uri, ex.getHttpCode());
			Logger.exception(message, ex);
			resp = ResponseFactory.exception(ex);
		} catch (Throwable ex) {
			Logger.exception(String.format("%s %s 500 internal error", method, uri), ex);
			resp = ResponseFactory.exception(ex);
		}

		writeResponse(exchange, resp);
	}

	private void writeResponse(HttpExchange exchange, Response resp) throws IOException {
		var headers = resp.getHeaders();
		for (var key : headers.keySet()) {
			var values = headers.get(key);
			for (var value: values) {
				exchange.getResponseHeaders().add(key, value);
			}
		}

		var body = resp.getBody();
		exchange.getResponseHeaders().set("Content-Type", resp.getContentType());
		exchange.sendResponseHeaders(resp.getHttpCode(), body.length);
		exchange.getResponseBody().write(body);
        exchange.getResponseBody().close();
	}
	
}

File: framework/kotlin/src/main/kotlin/soda/kotlin/web/BaseHandler.kt
package soda.kotlin.web

import com.sun.net.httpserver.HttpExchange
import com.sun.net.httpserver.HttpHandler
import soda.kotlin.web.exception.ServiceException
import soda.kotlin.web.resp.Response
import soda.kotlin.web.resp.ResponseFactory
import java.nio.charset.StandardCharsets

abstract class BaseHandler : HttpHandler {

    protected open fun doGet(exchange: HttpExchange?): Response {
        throw RuntimeException("not implemented")
    }

    protected open fun doPost(exchange: HttpExchange?): Response {
        throw RuntimeException("not implemented")
    }

    override fun handle(exchange: HttpExchange?) {
        if (exchange == null) {
            throw IllegalArgumentException("exchange may not be null")
        }

        val method = exchange.requestMethod
        val uri = exchange.requestURI
        var resp: Response

        try {
            val startNano = System.nanoTime()
            if (method == "GET") {
                resp = doGet(exchange)
            } else if (method == "POST") {
                resp = doPost(exchange)
            } else {
                throw RuntimeException("method $method not supported")
            }
            val endNano = System.nanoTime()
            val elapseMs = (endNano - startNano) / 1e6
            val code = resp.getHttpCode()
            Logger.info("$method $uri $code $elapseMs")
        } catch (ex: ServiceException) {
            val message = "$method $uri ${ex.httpCode} business error"
            Logger.exception(message, ex)
            resp = ResponseFactory.exception(ex)
        } catch (ex: Throwable) {
            val message = "$method $uri 500 internal error"
            Logger.exception(message, ex)
            resp = ResponseFactory.exception(ex)
        }

        writeResponse(exchange, resp)
    }

    private fun writeResponse(exchange: HttpExchange, resp: Response) {
        try {
            val headers = resp.getHeaders()
            for (key in headers.keys) {
                val values = headers[key]!!
                for (value in values) {
                    exchange.responseHeaders.add(key, value)
                }
            }

            val body = resp.getBody()
            exchange.responseHeaders.set("Content-Type", resp.getContentType())
            exchange.sendResponseHeaders(resp.getHttpCode(), body.size.toLong())

            exchange.responseBody.use {
                it.write(body)
            }
        } catch (ex: Exception) {
            Logger.exception("response error", ex)
        }
    }

}